modules/server/src/network/types.ts

```diff
@@ -0,0 +1,13 @@
+import { GraphQLSchema } from 'graphql';
+
+// environment config
+export type NetworkAggregationConfigInput = {
+	graphqlUrl: string;
+	documentType: string;
+	displayName: string;
+};
+
+// additional properties after initialising network
+export type NetworkAggregationConfig = NetworkAggregationConfigInput & {
+	schema: GraphQLSchema | null;
```

Comment:
have a feeling Jon won't like `null` (I'm a fan of it, though)

Reply:
I feel the same way. In this case I think it makes sense.

We've tried and failed to get a schema for a remote connection.
Errors have been caught and logged, but aren't fatal, doesn't stop other remote connections.
We still want to know that we have this connection in our env, we tried to connect + process, but no schema (versus processing/attempts still to complete ie. undefined)

Reply:
sound precisely as my own line of thought: I like using `null` as a deliberate value to differentiate from just "missing".

Reply:
In the process of validating `undefined` vs `null` you are both discussing states. Specifically I have seen the following states described:
- schema available (has type `GraphQLSchema`)
- schema not available (has type `null`)
- an attempt to fetch the schema has not been completed (has type `undefined`)

So using this logic, we are encoding stateful information about our fetch history into the `NetworkAggregationConfig.schema` type. When future devs interact with this type, none of this state information is being communicated, they just know what types the property could be. If we want to convey important state information I would suggest expanding your type to include either a `state` property, and maybe even use differentiated unions (see below).

More immediately relevant here though is that we don't have 3 states, we only have 2 - value present and value missing. TS has a built in convention for optional properties: `schema?: GraphQLSchema`. For this reason alone, I would encourage using undefined by default unless there is a need to introduce `null` into the code - the most common reason is that "external API uses null types" - this is relevant for GQL interfaces, but not here.

### Example
Example for encoding many states with differentiated union, not saying we need this:
```ts
type RemoteNodeSchemaAvailable = {
  state: 'available',
  schema: GraphQLSchema
}

type RemoteNodeSchemaUnavailable = {
  state: 'unavailable',
  reason: string
}

type RemoteNodeSchemaFetching = {
  state: 'fetching',
  start: Date
  timeout: Date
}

type RemoteNodeSchema = RemoteNodeSchemaAvailable | RemoteNodeSchemaUnavailable | RemoteNodeSchemaFetching;

type NetworkAggregationConfig = NetworkAggregationConfigInput & RemoteNodeSchema;
```

Reply:
One last thought here: there isn't much practical difference in most case. Code that consumes this type is mostly going to do like:

```ts
if(config.schema) {
  // do stuff with schema
} else {
  // handle case without it
}
```

or otherwise filter out the missing case. If this is your expectation, what have we added to our code base by mixing the use of `undefined` and `null`.